### ISSUE TITLE
UART interrupts not triggered on ARC

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#484]
**_Reported by ravi hegde:_**

Configuration
==========
<Arch> ARC
<Board>quark se devboard
<Kernel> nano
Bug detailed description
UART sample application does not receive interrupts when it is running on ARC core. 

Steps to reproduce:
1) Start with the UART application available in zephyr/samples/drivers/UART. Add the following to prj.conf
CONFIG_UART_QMSI=y
CONFIG_UART_INTERRUPT_DRIVEN=y

2) Build the UART application for ARC as "make BOARD=quark_se_sss_devboard"
3) Flash the application 

When the application starts running we see the poll mode works however the interrupt mode does not work. The same application works in interrupt mode when run on x86.

(Imported from Jira ZEP-546)

**Comment by nashif** [comment]
**_by Kuo-Lang Tseng:_**

ARC side does not have UART support so wondering if this should be expected behavior.  

(BTW, If you need to output message from ARC to serial, it needs to go thru IPM.)

**Comment by nashif** [comment]
**_by ravi hegde:_**

Hi,
We are using the UART that is on the common fabric and shared between x86 and ARC so both interrupt mode and polled mode should be supported. Infact, polled mode is working fine.
Also, regarding IPM-> we have printed messages without going through IPM. Ofcourse if you have app printing from x86 as well, then the characters are jumbled.

**Comment by nashif** [comment]
**_by Kuo-Lang Tseng:_**

As far as we know, the existing UART driver supports only for x86 side. It may be enhanced to support ARC also and this will require re-routing interrupt to ARC, looks like this is possible based on the data sheet. We have not planned that support but can add this plan for future version. Keep in mind after routing to ARC, LMT will not be able to use the same UART.

**Comment by nashif** [comment]
**_by ravi hegde:_**

fyi, the bare metal QMSI already supports UART interrupt routing to ARC. 
Zephyr does too for other driver such as SPI. So wondering if this is a bug or currently unsupported feature.

**Comment by nashif** [comment]
**_by Jesus Sanchez-Palencia:_**

Looks like a unsupported feature, _ravi hegde_ . _Kuo-Lang Tseng_ , do we have any plans to add this support?
_ravi hegde_  also, please note that if what you need is console output over UART for ARC, you can accomplish that in Zephyr through the Console IPM driver (it uses mailbox).

This is the current solution for both cores to get console output through the same UART.

**Comment by nashif** [comment]
**_by Kuo-Lang Tseng:_**

Afaik, there is currently no plan to support this. We can add a Jira task to add this, if this is needed. Yes, the IPM is current solution as pointed out by Jesus.

**Comment by nashif** [comment]
**_by Kuo-Lang Tseng:_**

Suggest to change the issue type as a task, instead of a bug, as this is not a bug but an enhancement?

**Comment by nashif** [comment]
**_by Rahul Singh:_**

We don't need console output, we need to send data over UART from ARC to other system.

**Comment by nashif** [comment]
**_by ravi hegde:_**

I can't seem to move this to a task. Should it be a requirement?

**Comment by nashif** [comment]
**_by Jesus Sanchez-Palencia:_**

So you want an UART driver for ARC, that is the requirement, right?

Meanwhile, I believe you can use mailbox (IPM) to send the data to Lakemont so it sends over UART.

**Comment by nashif** [comment]
**_by Kuo-Lang Tseng:_**

_ravi hegde_  If the bug cannot be changed to a requirement, can you close it and open a new one (requirement) so the current one will not get scrubbed by bug scrub process. Thanks.

**Comment by nashif** [comment]
**_by Kuo-Lang Tseng:_**

Hi Ravi, please change this bug to requirement, or open a new Jira for requirement and close this current one. Thanks.

**Comment by nashif** [comment]
**_by Mark Linkmeyer:_**

Hi _ravi hegde_ , I've assigned this bug to you to either convert it to a Story or open a new Story as _Kuo-Lang Tseng_  requests above.  Note that I clarified Kuo's request to say 'Story' and not 'Requirement'.  The Zephyr project doesn't have a Requirement issue type anymore because Stories are essentially the same thing and there's no need for both.  If you have any troubles converting it to a Story, please let me know. If you end up creating a new Story please make sure you link it to this bug as a 'related' issue, then close this bug.  Thx.

**Comment by nashif** [comment]
**_by ravi hegde:_**

Moved to a story. Do I need to reopen it?

**Comment by nashif** [comment]
**_by Kuo-Lang Tseng:_**

_ravi hegde_  Thanks. Yes, I just re-opened it.

**Comment by nashif** [comment]
**_by Anas Nashif:_**

UART is supported on ARC exactly the same way it is supported on X86, with the same driver. This has been a feature since 1.1 I guess. 
Make sure you disable IPM on ARC (which is enable by default) and probably you need some other configuration options to get this working, but it is 100% supported, in fact, you can have x86 print to uart0 and arc print to uart1 if configured correctly.

**Comment by nashif** [comment]
**_by Rahul Singh:_**

Hello Anas,

We have tested the UART sample application with IPM disabled. Poll mode works fine ,but interrupt mode is not working on ARC.
Can you please help us if something we are missing in config file because of that UART interrupt is not working on ARC.

Configuration used :
CONFIG_SERIAL=y
CONFIG_UART_INTERRUPT_DRIVEN=y
CONFIG_UART_CONSOLE=y
CONFIG_UART_QMSI=y
CONFIG_STDOUT_CONSOLE=y
CONFIG_IPM=n
CONFIG_IPM_QUARK_SE=n
CONFIG_IPM_QUARK_SE_MASTER=n
CONFIG_IPM_CONSOLE_RECEIVER=n
CONFIG_IPM_CONSOLE_SENDER=n

Regards,
Rahul

**Comment by nashif** [comment]
**_by Anas Nashif:_**

yes, I can confirm the interrupts are not working correctly, to verify this I use the shell sample application.

**Comment by nashif** [comment]
**_by Anas Nashif:_**

this should be moved to a bug, however Jira is not allowing me to move from story->bug.

**Comment by nashif** [comment]
**_by Anas Nashif:_**

there is another way to test this: samples/drivers/uart

this partially works, then the application hangs.

**Comment by nashif** [comment]
**_by Andy Ross:_**

Got it.  Basically the code was doing all the right stuff, but didn't realize that IRQ numbering and interrupt routing mask bits change between SS and IA.  The fix patch (and a few preceding cleanup bits) are up at:

```
remote:   https://gerrit.zephyrproject.org/r/4474 uart_console: Fix line endings
remote:   https://gerrit.zephyrproject.org/r/4475 samples/drivers/uart: Fix line endings
remote:   https://gerrit.zephyrproject.org/r/4476 samples/uart: More boards
remote:   https://gerrit.zephyrproject.org/r/4477 uart_qmsi: Get the interrupt handling right on ARC
```

**Comment by nashif** [comment]
**_by Andy Ross:_**

Mark fixed/resolved.  Hopefully I'm following the right process here.

**Comment by nashif** [comment]
**_by Kuo-Lang Tseng:_**

_Andy Ross_  Thanks, Almost right:) except that current workflow is to keep the status to Accepted (means in progress) and change to Resolved once patch is merged. (Not that obvious with current status values we have and _Mark Linkmeyer_  is working on getting the workflow updated to make it clearer).

**Comment by nashif** [comment]
**_by ethan gao:_**

It's verified that the interrupt mode on ARC was enabled and works now

**Comment by nashif** [comment]
Related to GH-401

**Comment by nashif** [comment]
Related to GH-601
